### PR TITLE
Issues(new) to develop

### PR DESCRIPTION
**Pull request by grogou**
Added Discord notification functionality

### CODE REVIEW

File: assets/js/admin.js [hunk 1]
```diff
@@ -24,66 +24,48 @@
...
-		$('.btm-users-settings').select2({
-			placeholder: 'Select users'
-		});
-	});
 
-	//custom bulk
-	$( '.btm-bulk-delete' ).click( function(){
-		if($(this).prop("checked") === true){
-			$( '.btm-delete' ).prop("checked", true);
+		if( $('#failed').attr('checked') ){
```
Comment (arsendovlatyan) [inline assets/js/admin.js]: @eriknovembit I'm not CSS guru, but I think working with ID `#failed` can bring up unexpected results ))

File: core/admin_menu/class-btm-admin-page-notification-single.php [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	public function __construct() {
+		$this->add_submenu_page();
+	}
+
+	// region Page
+
+	protected function add_submenu_page() {
+		$page_hook = add_submenu_page(
+			'admin.php',
+			__( 'Notification Rule', 'background_task_manager' ),
```
Comment (arsendovlatyan) [inline core/admin_menu/class-btm-admin-page-notification-single.php:24]: @eriknovembit let's just call it `Notifications`

Comment (grogou) [reply core/admin_menu/class-btm-admin-page-notification-single.php:24]: I fix it

File: assets/js/admin.js (hunk 1 above)
Comment (grogou) [reply assets/js/admin.js]: I think you right. I'm fixed them

File: data-provider/class-btm-notification-dao.php [hunk 3]
```diff
@@ -38,137 +38,107 @@ public function get_callbacks_table_name(){
...
-	 * @param I_BTM_Task $task
+	 * @param $id
+	 * @param $callback
+	 * @param $webhook
+	 * @param $report_type
 	 *
 	 * @return bool
 	 */
-	public function update( I_BTM_Task $task ){
+	public function update( $id, $callback, $webhook, $report_type ){
```
Comment (zzhovo) [inline data-provider/class-btm-notification-dao.php:127]: Wrap the arguments into a model, all the values coming from the front should be validated inside the model, refactor everywhere.

File: data-provider/class-btm-notification-dao.php [hunk 4]
```diff
@@ -38,137 +38,107 @@ public function get_callbacks_table_name(){
...
 		global $wpdb;
 
-		$callback_arguments = serialize( $task->get_callback_arguments() );
 		$updated = $wpdb->update(
-			$this->get_table_name(),
+			$this->get_callbacks_table_name(),
```
Comment (zzhovo) [inline data-provider/class-btm-notification-dao.php]: Rename this method to be get_table_name.

File: data-provider/class-btm-task-dao.php [hunk 5]
```diff
@@ -210,6 +210,78 @@ public function get_next_task_to_run(){
 		return $this->create_task_from_db_obj( $task_obj );
 	}
 
+	/**
+	 * @param int $hour
+	 *
+	 * @return array|bool
```
Comment (zzhovo) [inline data-provider/class-btm-task-dao.php:216]: Fix the return type.

File: core/class-btm-notification-runner.php [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	 * @param string $report_range
+	 *
+	 * @return bool
+	 */
+	public function report( $report_range ) {
+		$notifications = BTM_Notification_Dao::get_instance()->get_notification_rules();
+
+		foreach ( $notifications as $item ) {
+			$this->webhook = $item->webhook;
+			$report_type   = maybe_unserialize( $item->report_type );
```
Comment (zzhovo) [inline core/class-btm-notification-runner.php:124]: Do not use `maybe_unserialize`, keep a straight approach.

File: core/class-btm-notification-runner.php [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+			if ( in_array( $report_range, $report_type ) ) {
+				$hour = 0;
```
Comment (zzhovo) [inline core/class-btm-notification-runner.php:126]: Validate the `$report_range` and remove this line.

Comment (zzhovo) [reply core/class-btm-notification-runner.php:126]: Create an enum for report ranges.

File: core/class-btm-notification-runner.php [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+			}
+
+		}
+	}
+
+	/**
+	 * @param $task_callback_action
+	 * @param $status
+	 */
+	private function notification_form( $task_callback_action, $status ){
```
Comment (zzhovo) [inline core/class-btm-notification-runner.php:72]: `notification_form` is not a method name. 
The class that sends notifications should be singleton and stateless.

File: core/class-btm-notification-runner.php [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+	/**
+	 * BTM_Notification_Runner constructor.
+	 */
+	public function __construct() {
+	}
+
+	/**
+	 * @param I_BTM_Task $task
+	 */
+	public function notify_failed_task( $task ) {
```
Comment (zzhovo) [inline core/class-btm-notification-runner.php:47]: Do we have somewhere dedicated methods to send succeeded notifications?

File: core/class-btm-notification-runner.php [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	/**
+	 * @param $task_callback_action
+	 * @param $status
+	 */
+	private function notification_form( $task_callback_action, $status ){
+		$this->callback_action = $task_callback_action;
+		$this->status          = $status;
+		$this->task_url        = admin_url( 'admin.php?page=btm-task-view&action=edit&task_id=' . $this->task_id );
```
Comment (zzhovo) [inline core/class-btm-notification-runner.php:75]: `btm-task-view` is hardcoded.

File: core/cron-job/class-btm-cron-job-notification-daily.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+		return 'daily';
+	}
+
+	/**
+	 * The callback job, should not be called directly
+	 * Report daily
+	 */
+	public function on_cron_job_runs() {
+
+		BTM_Notification_Daily_Manager::get_instance();
```
Comment (zzhovo) [inline core/cron-job/class-btm-cron-job-notification-daily.php]: Remove this line.

File: core/cron-job/class-btm-cron-job-notification-hourly.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+		return 'hourly';
+	}
+
+	/**
+	 * The callback job, should not be called directly
+	 * Report hourly
+	 */
+	public function on_cron_job_runs() {
+
+		BTM_Notification_Hourly_Manager::get_instance();
```
Comment (zzhovo) [inline core/cron-job/class-btm-cron-job-notification-hourly.php]: Remove this line.

File: core/class-btm-notification-daily-manager.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	public function on_btm_task_notification_daily_report( BTM_Task_Run_Filter_Log $task_run_filter_log ){
+
+		$notification = new BTM_Notification_Runner();
+		$reported = $notification->report( $this->get_report_range() );
+		if( $reported === true ){
+			$task_run_filter_log->add_log( __( 'The daily report is done', 'background_task_manager' ) );
+			$task_run_filter_log->set_failed( false );
+		}else{
+			$task_run_filter_log->add_log( __( 'Nothing to report', 'background_task_manager' ) );
+			$task_run_filter_log->set_failed( true );
```
Comment (zzhovo) [inline core/class-btm-notification-daily-manager.php]: If there's nothing to report the task should not be considered as failed.

File: core/class-btm-plugin-options.php [hunk 14]
```diff
@@ -514,4 +515,51 @@ public function update_delete_old_entities_cron_job_interval_in_days( $interval
 
 		return false;
 	}
+
+
+	/**
+	 * @var string
+	 */
+	private $show_system;
```
Comment (zzhovo) [inline core/class-btm-plugin-options.php:523]: Value should be bool.

File: core/class-btm-notification-hourly-manager.php [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+
+if ( ! defined( 'BTM_PLUGIN_ACTIVE' ) ) {
+	exit;
+}
+
+/**
+ * Class BTM_Notification_Hourly_Manager
+ */
+class BTM_Notification_Hourly_Manager {
```
Comment (zzhovo) [inline core/class-btm-notification-hourly-manager.php]: There should be only one manager and one btm task, the time range(daily, hourly) should be its callback argument.

Comment (arsendovlatyan) [reply core/class-btm-notification-hourly-manager.php]: @eriktad բա որ ասում էի )))

File: core/class-btm-notification-runner.php (hunk 8 above)
Comment (grogou) [reply core/class-btm-notification-runner.php:72]: this method only creates discord notification form, but don't send notification

File: core/class-btm-notification-runner.php (hunk 9 above)
Comment (grogou) [reply core/class-btm-notification-runner.php:47]: No, I think we don't need succeeding notification functionality

File: core/class-btm-notification-daily-manager.php (hunk 13 above)
Comment (grogou) [reply core/class-btm-notification-daily-manager.php]: why? ))

File: core/class-btm-notification-hourly-manager.php (hunk 15 above)
Comment (grogou) [reply core/class-btm-notification-hourly-manager.php]: It's done )))

**Pull request merged**